src/examples/libpmemobj/hashmap/hashmap_rp.c

```diff
@@ -0,0 +1,700 @@
+		args.pos = increment_pos(hashmap, args.pos);
+		dist += 1;
+	}
+	fprintf(stderr, "insertion requires too many swaps\n");
+	pmemobj_cancel(pop, args.actv, args.actv_cnt);
+
+	return -1;
+}
+
+/*
+ * index_lookup -- checks if given key exists in hashmap.
+ * Returns index number if key was found, 0 otherwise.
+ */
+static uint64_t
+index_lookup(const struct hashmap_rp *hashmap, uint64_t key)
+{
+	const uint64_t hash_lookup = hash(hashmap, key);
+	uint64_t pos = hash_lookup;
+	uint64_t dist = 0;
+
+	const struct entry *entry_p = NULL;
+	while (1) {
+		entry_p = (struct entry *)pmemobj_direct(hashmap->entries.oid);
+		entry_p += pos;
+
+		if (entry_p->hash == 0 || dist > probe_distance(hashmap,
+							entry_p->hash, pos))
+			return 0;
+		else if (entry_p->hash == hash_lookup && entry_p->key == key)
+			return pos;
+
+		pos = increment_pos(hashmap, pos);
+		dist++;
+	}
+}
+
+/*
+ * entries_cache -- cache entries from first argument in entries from second
+ * argument
+ */
+static int
+entries_cache(PMEMobjpool *pop, const struct hashmap_rp *from,
+	struct hashmap_rp *to)
+{
+	struct entry *current_el =
+			(struct entry *)pmemobj_direct(from->entries.oid);
+	for (int i = 0; i < from->capacity; ++i, ++current_el) {
+		if (entry_is_empty(current_el->hash))
+			continue;
+
+		if (insert_helper(pop, to, current_el->key,
+				current_el->value) == -1)
+			return -1;
+	}
+	assert(from->count == to->count);
+
+	return 0;
+}
+
+/*
+ * hm_rp_rebuild -- rebuilds the hashmap with a new capacity.
+ * Returns 0 on success, -1 otherwise.
+ */
+static int
+hm_rp_rebuild(PMEMobjpool *pop, TOID(struct hashmap_rp) hashmap,
+		size_t capacity_new)
+{
+	/*
+	 * We will need 6 actions:
+	 * - 1 action to set new capacity
+	 * - 1 action to set new resize threshold
+	 * - 1 action to alloc memory for new entries
+	 * - 1 action to free old entries
+	 * - 2 actions to set new oid pointing to new entries
+	 */
+	struct pobj_action actv[6];
+	size_t actv_cnt = 0;
+
+	size_t sz_alloc = sizeof(struct entry) * capacity_new;
+	uint64_t resize_threshold_new = (uint64_t)(capacity_new *
+		HASHMAP_RP_LOAD_FACTOR);
+
+	pmemobj_set_value(pop, &actv[actv_cnt++], &D_RW(hashmap)->capacity,
+		capacity_new);
+
+	pmemobj_set_value(pop, &actv[actv_cnt++],
+		&D_RW(hashmap)->resize_threshold, resize_threshold_new);
+
+	struct hashmap_rp hashmap_rebuild;
+	hashmap_rebuild.count = 0;
+	hashmap_rebuild.capacity = capacity_new;
+	hashmap_rebuild.resize_threshold = resize_threshold_new;
+	hashmap_rebuild.entries = POBJ_XRESERVE_ALLOC(pop, struct entry,
+						sz_alloc, &actv[actv_cnt],
+						POBJ_XALLOC_ZERO);
+
+	if (TOID_IS_NULL(hashmap_rebuild.entries)) {
+		fprintf(stderr, "hashmap rebuild failed: %s\n",
+			pmemobj_errormsg());
+		goto rebuild_err;
+	}
+	actv_cnt++;
+
+	if (hashmap_rp_debug) {
+		free(swaps_array);
+		swaps_array = (int *)calloc(capacity_new, sizeof(int));
+		if (!swaps_array)
+			goto rebuild_err;
+	}
+
+	if (entries_cache(pop, D_RW(hashmap), &hashmap_rebuild) == -1)
```

Comment:
This is sub-optimal. You are using redo logs for reserved memory, which you don't have to do. I'm afraid you will need to re-implement inserting for rebuilding.